janusgraph-core/src/main/java/org/janusgraph/graphdb/util/MultiDistinctUnorderedIterator.java

```diff
@@ -0,0 +1,63 @@
+// Copyright 2020 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.util;
+
+import org.apache.tinkerpop.gremlin.structure.util.CloseableIterator;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+public class MultiDistinctUnorderedIterator<E> extends CloseableAbstractIterator<E> {
+
+    private final Set<E> allElements = new HashSet<E>();
+    private final CloseableIterator<E> iterator;
+    private final int limit;
+    private long count;
+
+    public MultiDistinctUnorderedIterator(final int lowLimit, final int highLimit, final List<CloseableIterator<E>> iterators) {
+        Objects.requireNonNull(iterators);
+        iterator = IteratorUtils.concat(iterators);
+        limit = highLimit;
+
+        long i = 0;
+        while (i < lowLimit && hasNext()) {
+            next();
+            i++;
+        }
+    }
+
+    @Override
+    protected E computeNext() {
+        if (count < limit) {
+            while (iterator.hasNext()) {
+                E elem = iterator.next();
+                if (!allElements.contains(elem)) {
```

Comment:
Just an idea: For large sets it can be quite inefficient to perform this check, because hash sets are not cache friendly at all. So before this check, one could query a bloom filter. If the filter says that this element is new, the check can be skipped.
On the downside, we would still need to collect the elements in `allElements` so insertions would still consume as much time as before. Therefore I'm not sure how significant a change like this would be.

This is just a general consideration and has nothing to do with your PR. So feel free to dismiss it :smile:

Reply:
Sounds interesting. My intuition is that would be even slower, but worth giving it a try. If I got some time I'll do some experiments and probably open up a new issue.

Reply:
Also, a small nitpick, why not to add `elem` to the set instead of checking for its existence and then adding it to the set?
I.e. the next code is 1 line shorted and doesn't call `contains` operation:
```
if(allElements.add(elem)){
  count++;
  return elem;
}
```
That said, I didn't check if bloom filter is actually faster for this logic or not. Not sure about that

Reply:
@porunov With your idea implemented, I'm now pretty sure a bloom filter would just cause overhead here. Even with one, either `contains` or `add` would be called anyway, so I don't see an improvement there anymore.